sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/aio/_sync_token_async.py

```diff
@@ -0,0 +1,94 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+from typing import Any, Dict
+from asyncio import Lock
+from .._sync_token import SyncToken
+from azure.core.pipeline import PipelineRequest, PipelineResponse
+from azure.core.pipeline.policies import SansIOHTTPPolicy
+
+
+class AsyncSyncTokenPolicy(SansIOHTTPPolicy):
+    """A simple policy that enable the given callback
+    with the response.
+    :keyword callback raw_response_hook: Callback function. Will be invoked on response.
+    """
+
+    def __init__(self, **kwargs):  # pylint: disable=unused-argument
+        # type: (**Any) -> None
+        self._sync_token_header = "Sync-Token"
+        self._sync_tokens = {}  # type: Dict[str, Any]
+        self._lock = Lock()
+
+    async def on_request(self, request):  # type: ignore # pylint: disable=arguments-differ
+        # type: (PipelineRequest) -> None
+        """This is executed before sending the request to the next policy.
+        :param request: The PipelineRequest object.
+        :type request: ~azure.core.pipeline.PipelineRequest
+        """
+        async with self._lock:
+            sync_token_header = ",".join(str(x) for x in self._sync_tokens.values())
+            if sync_token_header:
+                request.http_request.headers.update(
+                    {self._sync_token_header: sync_token_header}
+                )
+
+    async def on_response(self, request, response):  # type: ignore # pylint: disable=arguments-differ
+        # type: (PipelineRequest, PipelineResponse) -> None
+        """This is executed after the request comes back from the policy.
+        :param request: The PipelineRequest object.
+        :type request: ~azure.core.pipeline.PipelineRequest
+        :param response: The PipelineResponse object.
+        :type response: ~azure.core.pipeline.PipelineResponse
+        """
+        sync_token_header = response.http_response.headers.get(self._sync_token_header)
+        if not sync_token_header:
+            return
+        sync_token_strings = sync_token_header.split(",")
+        if not sync_token_strings:
+            return
+        async with self._lock:
+            for sync_token_string in sync_token_strings:
+                sync_token = SyncToken.from_sync_token_string(sync_token_string)
+                self._update_sync_token(sync_token)
+
+    async def add_token(self, full_raw_tokens):
+        # type: (str) -> None
+        async with self._lock:
+            raw_tokens = full_raw_tokens.split(",")
+            for raw_token in raw_tokens:
+                sync_token = SyncToken.from_sync_token_string(raw_token)
+                self._update_sync_token(sync_token)
+
+    def _update_sync_token(self, sync_token):
```

Comment:
Should we use lock in this method rather than every caller need to handle the lock?

Reply:
By "caller" do you mean the `update_sync_token` caller or something different?

Reply:
I mean maybe it is better to have 
get_lock() in update_sync_token than having it in both 
add_token() & on_response()?